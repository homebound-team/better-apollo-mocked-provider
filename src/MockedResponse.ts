import { FetchResult, GraphQLRequest } from "@apollo/client";

export type ResultFunction<T> = () => T;

export interface MockedResponse<TData = Record<string, any>> {
  request: GraphQLRequest;
  result?: FetchResult<TData> | ResultFunction<FetchResult<TData>>;
  error?: Error;
  delay?: number;
  // Homebound note: Added as part of making MockLink behavior more sane/less opaque.
  loading?: boolean;
  newData?: ResultFunction<FetchResult>;
  requested?: number;
}
