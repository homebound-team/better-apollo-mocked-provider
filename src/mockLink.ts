import { ApolloLink, FetchResult, GraphQLRequest, Operation } from "@apollo/client";
import {
  addTypenameToDocument,
  Observable,
  removeClientSetsFromDocument,
  removeConnectionDirectiveFromDocument,
} from "@apollo/client/utilities";
import stringify from "fast-json-stable-stringify";
import { print } from "graphql/language/printer";
import { invariant } from "ts-invariant";
import { MockedResponse, ResultFunction } from "./MockedResponse";

/**
 * A fork of apollo-client's `MockLink` to be more developer friendly.
 *
 * The biggest change is being more chatty on failures so that developers
 * can more easily see "ah that is why this didn't match".
 */
export class MockLink extends ApolloLink {
  public operation: Operation | undefined;
  public addTypename: Boolean = true;
  private mockedResponsesByKey: { [key: string]: MockedResponse[] } = {};
  private madeResponsesByKey: { [key: string]: MockedResponse[] } = {};

  constructor(mockedResponses: ReadonlyArray<MockedResponse>, addTypename: Boolean = true) {
    super();
    this.addTypename = addTypename;
    mockedResponses.forEach((res) => this.addMockedResponse(res));
  }

  public addMockedResponse(mockedResponse: MockedResponse) {
    const normalizedMockedResponse = normalizeMockedResponse(mockedResponse);
    const key = requestToKey(normalizedMockedResponse.request, this.addTypename);
    let mockedResponses = this.mockedResponsesByKey[key];
    if (!mockedResponses) {
      mockedResponses = [];
      this.mockedResponsesByKey[key] = mockedResponses;
    }
    mockedResponses.push(normalizedMockedResponse);
  }

  public request(operation: Operation): Observable<FetchResult> {
    this.operation = operation;

    const key = requestToKey(operation, this.addTypename);

    // Homebound note: note that we use stringify to:
    //
    // a) strip undefined keys, and
    // b) call `.toJSON` on any not-actually-JSON values in the variables, i.e. user-defined types
    //
    // This is really just a bug fix that should be up-streamed, see:
    //
    // See https://github.com/apollographql/apollo-client/issues/6771
    const requestVariables = stringify(operation.variables || {});

    // Homebound note: little bit of refactoring here for simplification.
    const possibleRequests = this.mockedResponsesByKey[key] || [];
    const responseIndex = possibleRequests.findIndex((res) => {
      return requestVariables === stringify(res.request.variables || {});
    });

    if (responseIndex === -1) {
      const madeAny = !!this.madeResponsesByKey[key];
      const errorPrefix = madeAny ? "No more" : "No";
      const potentialVariables = possibleRequests.map((mr) => stringify(mr.request.variables || {}));
      const madeVariables = (this.madeResponsesByKey[key] || []).map((mr) => stringify(mr.request.variables || {}));

      // The apollo-client impl returns `null` from here. This "seems fine" but the super tricky
      // thing is that `this.onError` turns into `throw error` which is purposefully suppressed if
      // `first = true`, per the "blindly catching".
      //
      // https://github.com/apollographql/apollo-client/blob/master/src/core/ObservableQuery.ts#L553
      //
      // I believe that ObservableQuery comment is fine, however it seems this is how MockLink was
      // achieving the "leave un-mocked requests as loading". Because we don't return loading, we
      // throw and usually this is suppressed.
      //
      // _Unless_ this was not the 1st request, in which case the hard-coded "No more mocked responses"
      // error text actually makes sense.
      //
      // This all seems too tricky, so we're making a breaking change where instead of `return null`,
      // we're going to return the observableResult with error set.
      //
      // This means that clients must explicitly opt-in to the `stay loading behavior`, i.e. via
      // the MockedResponse.locking that we've added.
      const lines = [
        `${errorPrefix} mocked responses for the query: ${operation.operationName}`,
        `Actual variables:`,
        requestVariables,
        ``,
        `Mocked variables:`,
        ...potentialVariables,
        // If there were previous requests, let the developer know that we know about them, i.e. their setup is correct
        ...(madeVariables.length === 0 ? [] : ["Already made requests", ...madeVariables]),
      ];
      const message = lines.join("\n");
      // Highlight this as a usage error
      console.error(message);
      return observableResult(undefined, new Error(message), false, undefined);
    }

    const response = possibleRequests[responseIndex];

    if (response.requested === undefined) {
      response.requested = 0;
    }
    response.requested++;

    // Homebound note: This is a breaking change: we only remove the request
    // from possibleRequests if the user has either:
    //
    // a) explicitly provided multiple requests for this key, or
    // b) set a `newData` on 1st mocked response
    //
    // This is primarily for developer ergonomics, i.e. most other mocking libraries
    // will keep re-using the same request.
    const { newData } = response;
    if (possibleRequests.length > 1 || newData) {
      possibleRequests.splice(responseIndex, 1);
      if (newData) {
        response.result = newData();
        possibleRequests.push(response);
      }
      // Record that this request was made instead of throwing it away.
      let made = this.madeResponsesByKey[key];
      if (!made) {
        made = [];
        this.madeResponsesByKey[key] = made;
      }
      made.push(response);
    }

    const { result, error, loading, delay } = response;
    if (!result && !error && loading === undefined) {
      const message = `Mocked response should contain either result, error, or loading: ${key}`;
      // Highlight this as a usage error
      console.error(message);
      return observableResult(result, new Error(message), false, delay);
    }

    return observableResult(result, error, loading, delay);
  }
}

function observableResult(
  result: FetchResult | ResultFunction<FetchResult> | undefined,
  error: Error | undefined,
  loading: boolean | undefined,
  delay: number | undefined,
) {
  return new Observable<FetchResult>((observer) => {
    const resolve = () => {
      if (error) {
        observer.error(error);
      } else if (result) {
        observer.next(typeof result === "function" ? (result as ResultFunction<FetchResult>)() : result);
        observer.complete();
      } else if (loading) {
        // leave it loading
      }
    };
    const timer = setTimeout(resolve, delay ?? 0);
    return () => clearTimeout(timer);
  });
}

// Homebound note: Removed the cloneDeep so that the caller can observe our .requested assignment.
function normalizeMockedResponse(mockedResponse: MockedResponse): MockedResponse {
  const queryWithoutConnection = removeConnectionDirectiveFromDocument(mockedResponse.request.query);
  invariant(queryWithoutConnection, "query is required");
  mockedResponse.request.query = queryWithoutConnection!;
  const query = removeClientSetsFromDocument(mockedResponse.request.query);
  if (query) {
    mockedResponse.request.query = query;
  }
  return mockedResponse;
}

export interface MockApolloLink extends ApolloLink {
  operation?: Operation;
}

// Pass in multiple mocked responses, so that you can test flows that end up
// making multiple queries to the server.
// NOTE: The last arg can optionally be an `addTypename` arg.
export function mockSingleLink(...mockedResponses: Array<any>): MockApolloLink {
  // To pull off the potential typename. If this isn't a boolean, we'll just
  // set it true later.
  let maybeTypename = mockedResponses[mockedResponses.length - 1];
  let mocks = mockedResponses.slice(0, mockedResponses.length - 1);

  if (typeof maybeTypename !== "boolean") {
    mocks = mockedResponses;
    maybeTypename = true;
  }

  return new MockLink(mocks, maybeTypename);
}

function requestToKey(request: GraphQLRequest, addTypename: Boolean): string {
  const queryString = request.query && print(addTypename ? addTypenameToDocument(request.query) : request.query);
  const requestKey = { query: queryString };
  return JSON.stringify(requestKey);
}
