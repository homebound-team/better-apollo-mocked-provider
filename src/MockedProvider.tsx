import {
  ApolloCache,
  ApolloClient,
  ApolloLink,
  ApolloProvider,
  DefaultOptions,
  InMemoryCache,
  Resolvers,
} from "@apollo/client";
import React from "react";
import { MockedResponse } from "./MockedResponse";
import { MockLink } from "./mockLink";

// Homebound note: This is ~100% unchanged from apollo.
export interface MockedProviderProps<TSerializedCache = {}> {
  mocks?: ReadonlyArray<MockedResponse>;
  addTypename?: boolean;
  defaultOptions?: DefaultOptions;
  cache?: ApolloCache<TSerializedCache>;
  resolvers?: Resolvers;
  childProps?: object;
  children?: React.ReactElement;
  link?: ApolloLink;
}

export interface MockedProviderState {
  client: ApolloClient<any>;
}

export class MockedProvider extends React.Component<MockedProviderProps, MockedProviderState> {
  public static defaultProps: MockedProviderProps = {
    addTypename: true,
  };

  constructor(props: MockedProviderProps) {
    super(props);

    const { mocks, addTypename, defaultOptions, cache, resolvers, link } = this.props;
    const client = new ApolloClient({
      cache: cache || new InMemoryCache({ addTypename }),
      defaultOptions,
      link: link || new MockLink(mocks || [], addTypename),
      resolvers,
    });

    this.state = { client };
  }

  public render() {
    const { children, childProps } = this.props;
    return children ? (
      <ApolloProvider client={this.state.client}>
        {React.cloneElement(React.Children.only(children), { ...childProps })}
      </ApolloProvider>
    ) : null;
  }

  public componentWillUnmount() {
    // Since this.state.client was created in the constructor, it's this
    // MockedProvider's responsibility to terminate it.
    this.state.client.stop();
  }
}
