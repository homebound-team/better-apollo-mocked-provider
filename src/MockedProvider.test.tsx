import { ApolloLink, InMemoryCache, useQuery } from "@apollo/client";
import { render, wait } from "@testing-library/react";
import { DocumentNode } from "graphql";
import gql from "graphql-tag";
import React from "react";
import { MockedProvider } from "./MockedProvider";
import { MockedResponse } from "./MockedResponse";
import { MockLink } from "./mockLink";

const variables = {
  username: "mock_username",
};

const userWithoutTypeName = {
  id: "user_id",
};

const user = {
  __typename: "User",
  ...userWithoutTypeName,
};

const query: DocumentNode = gql`
  query GetUser($username: String!) {
    user(username: $username) {
      id
    }
  }
`;

const queryWithTypename: DocumentNode = gql`
  query GetUser($username: String!) {
    user(username: $username) {
      id
      __typename
    }
  }
`;

const mocks: ReadonlyArray<MockedResponse> = [
  {
    request: {
      query,
      variables,
    },
    result: { data: { user } },
  },
];

interface Data {
  user: {
    id: string;
  };
}

interface Variables {
  username: string;
}

let errorThrown = false;
const errorLink = new ApolloLink((operation, forward) => {
  // Homebound note: We've changed this from the default MockedProvider.test.tsx
  // setup b/c we no longer throw errors from the MockLink impl (see longer note
  // in that file). Instead we always use the observableResult + a console.error
  // to let the programmer know what's going on.
  //
  // Normally I would not prefer this, but per note in MockLink link, the 1st error
  // thrown from links is suppressed "because they show up later" (...except that
  // they don't for MockLink's prior behavior).
  const observer = forward(operation);
  observer.subscribe(
    () => {},
    () => {
      errorThrown = true;
    },
  );
  return observer;
});

describe("MockedProvider", () => {
  beforeEach(() => {
    errorThrown = false;
  });

  it.async("should mock the data", async (resolve, reject) => {
    function Component({ username }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data!.user).toMatchSnapshot();
      }
      return null;
    }

    const mock: MockedResponse = {
      request: { query, variables },
      result: { data: { user } },
    };

    render(
      <MockedProvider mocks={[mock]}>
        <Component {...variables} />
      </MockedProvider>,
    );

    return wait(() => {
      expect(mock.requestedCount).toEqual(1);
    }).then(resolve, reject);
  });

  it.async("should allow querying with the typename", async (resolve, reject) => {
    function Component({ username }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, {
        variables,
      });
      if (!loading) {
        expect(data!.user).toMatchSnapshot();
      }
      return null;
    }

    const mocksWithTypename = [
      {
        request: {
          query: queryWithTypename,
          variables,
        },
        result: { data: { user } },
      },
    ];

    render(
      <MockedProvider mocks={mocksWithTypename}>
        <Component {...variables} />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });

  it.async("should allow using a custom cache", async (resolve, reject) => {
    const cache = new InMemoryCache();
    cache.writeQuery({
      query,
      variables,
      data: { user },
    });

    function Component({ username }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data).toMatchObject({ user });
      }
      return null;
    }

    render(
      <MockedProvider mocks={[]} cache={cache}>
        <Component {...variables} />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });

  it.async("should error if the variables in the mock and component do not match", async (resolve, reject) => {
    function Component({ ...variables }: Variables) {
      useQuery<Data, Variables>(query, { variables });
      return null;
    }

    const variables2 = {
      username: "other_user",
    };

    const link = ApolloLink.from([errorLink, new MockLink(mocks)]);

    render(
      <MockedProvider link={link}>
        <Component {...variables2} />
      </MockedProvider>,
    );

    return wait(() => {
      expect(errorThrown).toBeTruthy();
    }).then(resolve, reject);
  });

  it.async("should error if the variables do not deep equal", async (resolve, reject) => {
    function Component({ ...variables }: Variables) {
      useQuery<Data, Variables>(query, { variables });
      return null;
    }

    const mocks2 = [
      {
        request: {
          query,
          variables: {
            age: 13,
            username: "some_user",
          },
        },
        result: { data: { user } },
      },
    ];

    const variables2 = {
      username: "some_user",
      age: 42,
    };

    const link = ApolloLink.from([errorLink, new MockLink(mocks2)]);

    render(
      <MockedProvider link={link}>
        <Component {...variables2} />
      </MockedProvider>,
    );

    return wait(() => {
      expect(errorThrown).toBeTruthy();
    }).then(resolve, reject);
  });

  it.async("should not error if the variables match but have different order", async (resolve, reject) => {
    function Component({ ...variables }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, {
        variables,
      });
      if (!loading) {
        expect(data).toMatchSnapshot();
      }
      return null;
    }

    const mocks2 = [
      {
        request: {
          query,
          variables: {
            age: 13,
            username: "some_user",
          },
        },
        result: { data: { user } },
      },
    ];

    const variables2 = {
      username: "some_user",
      age: 13,
    };

    render(
      <MockedProvider mocks={mocks2}>
        <Component {...variables2} />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });

  it.async("should support mocking a network error", async (resolve, reject) => {
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, {
        variables,
      });
      if (!loading) {
        expect(error).toEqual(new Error("something went wrong"));
      }
      return null;
    }

    const mocksError = [
      {
        request: {
          query,
          variables,
        },
        error: new Error("something went wrong"),
      },
    ];

    render(
      <MockedProvider mocks={mocksError}>
        <Component {...variables} />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });

  it.async("should error if the query in the mock and component do not match", async (resolve, reject) => {
    function Component({ ...variables }: Variables) {
      useQuery<Data, Variables>(query, { variables });
      return null;
    }

    const mocksDifferentQuery = [
      {
        request: {
          query: gql`
            query OtherQuery {
              otherQuery {
                id
              }
            }
          `,
          variables,
        },
        result: { data: { user } },
      },
    ];

    const link = ApolloLink.from([errorLink, new MockLink(mocksDifferentQuery)]);

    render(
      <MockedProvider link={link}>
        <Component {...variables} />
      </MockedProvider>,
    );

    return wait(() => {
      expect(errorThrown).toBeTruthy();
    }).then(resolve, reject);
  });

  it("should pass down props prop in mock as props for the component", () => {
    function Component({ ...variables }) {
      expect(variables.foo).toBe("bar");
      expect(variables.baz).toBe("qux");
      return null;
    }

    render(
      <MockedProvider mocks={mocks} childProps={{ foo: "bar", baz: "qux" }}>
        <Component {...variables} />
      </MockedProvider>,
    );
  });

  it("should not crash on unmount if there is no query manager", () => {
    function Component() {
      return null;
    }

    const { unmount } = render(
      <MockedProvider>
        <Component />
      </MockedProvider>,
    );

    unmount();
  });

  it.async("should support returning mocked results from a function", async (resolve, reject) => {
    let resultReturned = false;

    const testUser = {
      __typename: "User",
      id: 12345,
    };

    function Component({ ...variables }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, {
        variables,
      });
      if (!loading) {
        expect(data!.user).toEqual(testUser);
        expect(resultReturned).toBe(true);
      }
      return null;
    }

    const testQuery: DocumentNode = gql`
      query GetUser($username: String!) {
        user(username: $username) {
          id
        }
      }
    `;

    const testVariables = {
      username: "jsmith",
    };
    const testMocks = [
      {
        request: {
          query: testQuery,
          variables: testVariables,
        },
        result() {
          resultReturned = true;
          return {
            data: {
              user: {
                __typename: "User",
                id: 12345,
              },
            },
          };
        },
      },
    ];

    render(
      <MockedProvider mocks={testMocks}>
        <Component {...testVariables} />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });
});

describe("@client testing", () => {
  it.async("should support @client fields with a custom cache", async (resolve, reject) => {
    const cache = new InMemoryCache();

    cache.writeQuery({
      query: gql`
        {
          networkStatus {
            isOnline
          }
        }
      `,
      data: {
        networkStatus: {
          __typename: "NetworkStatus",
          isOnline: true,
        },
      },
    });

    function Component() {
      const { loading, data } = useQuery(gql`
        {
          networkStatus @client {
            isOnline
          }
        }
      `);
      if (!loading) {
        expect(data!.networkStatus.__typename).toEqual("NetworkStatus");
        expect(data!.networkStatus.isOnline).toEqual(true);
      }
      return null;
    }

    render(
      <MockedProvider cache={cache}>
        <Component />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });

  it.async("should support @client fields with field policies", async (resolve, reject) => {
    const cache = new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            networkStatus() {
              return {
                __typename: "NetworkStatus",
                isOnline: true,
              };
            },
          },
        },
      },
    });

    function Component() {
      const { loading, data } = useQuery(gql`
        {
          networkStatus @client {
            isOnline
          }
        }
      `);
      if (!loading) {
        expect(data!.networkStatus.__typename).toEqual("NetworkStatus");
        expect(data!.networkStatus.isOnline).toEqual(true);
      }
      return null;
    }

    render(
      <MockedProvider cache={cache}>
        <Component />
      </MockedProvider>,
    );

    return wait().then(resolve, reject);
  });
});
