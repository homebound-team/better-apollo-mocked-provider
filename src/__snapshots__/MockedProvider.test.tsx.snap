// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MockedProvider should allow querying with the typename 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
}
`;

exports[`MockedProvider should not error if the variables match but have different order 1`] = `
Object {
  "user": Object {
    "__typename": "User",
    "id": "user_id",
  },
}
`;
